src: unload native addons when the environment is destroyed

gabrielschulhof opened the pull request:
This is an alternative to https://github.com/nodejs/node/pull/23319
which attaches the loaded addons to the environment and closes them
when the environment is destroyed.

It includes the test/addons/worker-addon,
test/addons/hello-world/test-worker.js and the documentation update from the original PR written
by @addaleax in a separate commit.

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

joyeecheung commented on test/addons/worker-addon/test.js:
```diff
@@ -0,0 +1,21 @@
+const child_process = require('child_process');
+const path = require('path');
+const { Worker } = require('worker_threads');
+const binding = path.resolve(__dirname, `./build/${common.buildType}/binding`);
```
Can you add a test when both the main thread and the worker thread require the same addon?

gabrielschulhof commented on src/node_binding.h:
```diff
@@ -46,6 +52,68 @@ extern bool node_is_initialized;
 
 namespace binding {
 
+class DLib {
```
@bnoordhuis I thought `inline` method bodies, if not right there in the class definition, need to also be in a header file, like a `-inl.h` file. Should I add one of those?

gabrielschulhof replied:
`std::list` does the trick. Don't need random access, after all.

gabrielschulhof commented on src/env-inl.h:
```diff
@@ -396,6 +396,15 @@ inline uv_loop_t* Environment::event_loop() const {
 
+inline void Environment::TryLoadAddon(const char* filename,
+                               int flags,
+                               std::function<bool(binding::DLib*)> was_loaded) {
```
@joyeecheung thanks! That helped 🙂

gabrielschulhof commented:
@joyeecheung I have addressed your review comments.

@bnoordhuis I have addressed your review comments. I put the body of `DLib` into a new `node_binding-inl.h' which is included from the bottom of `node_binding.h`.

bnoordhuis commented on src/node_binding-inl.h:
```diff
@@ -0,0 +1,57 @@
+#ifndef SRC_NODE_BINDING_INL_H_
+#define SRC_NODE_BINDING_INL_H_
+
```
`#include "node_binding.h"` here - the rule is that you can include the -inl.h variant without having to include the main header file as well.

Then, include this file only in the compilation units that actually use it.

That said, I recommend putting this in a .cc file instead of a header file because there's really no need to inline it in every compilation unit where it's used, that only leads to code bloat.

gabrielschulhof replied:
TBH I'd rather keep it `inline`. `Open()` is only called from `node_binding.cc` and `Close()` is only called from `node_binding.cc` and `env.cc` and, if we assume a future where many worker threads come and go the indirection of at least two function calls per thread per native addon may accumulate.

bnoordhuis replied:
That's not what `inline` does. It's not a compiler hint, it's a linkage directive; it means the function or method won't be visible outside the compilation unit.

That aside, two function calls aren't going to show up in profiles if you're also doing things like manipulating linked lists.

gabrielschulhof replied:
@bnoordhuis I've moved the definitions of the functions back into `src/node_binding.cc`.

bnoordhuis commented on src/node_binding.cc:
```diff
@@ -246,87 +174,92 @@ void DLOpen(const FunctionCallbackInfo<Value>& args) {
-    Local<String> errmsg = OneByteString(env->isolate(), dlib.errmsg_.c_str());
-    dlib.Close();
+  env->TryLoadAddon(*filename, flags, [&](DLib* dlib) {
+    bool is_opened = dlib->Open();
```
Suggestion: make it `const bool`

addaleax commented on src/env.cc:
```diff
@@ -267,6 +267,11 @@ Environment::~Environment() {
     TRACING_CATEGORY_NODE1(environment), "Environment", this);
+
+  // Dereference all addons that were loaded into this environment.
+  for (auto& addon : loaded_addons_) {
```
```suggestion
  for (binding::DLib& addon : loaded_addons_) {
```

gabrielschulhof commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/19417/

gabrielschulhof commented:
Failure looks legit.

gabrielschulhof commented:
Resume, just to see if it's 100% reproducible:

https://ci.nodejs.org/job/node-test-pull-request/19427/

addaleax commented:
Resume CI: https://ci.nodejs.org/job/node-test-pull-request/19569/

Trott commented:
Fresh CI rather than Resume Build so that it pulls in the fact that a bunch of tests are currently marked flaky: https://ci.nodejs.org/job/node-test-pull-request/19572/

Trott commented:
Any chance that this AIX failure is related?

```console
16:58:17 not ok 2208 addons/at-exit/test
16:58:17   ---
16:58:17   duration_ms: 0.723
16:58:17   severity: crashed
16:58:17   exitcode: -4
16:58:17   stack: |-
16:58:17   ...
```

I'm guessing not, but you tell me. ¯\\_(ツ)_/¯

In the meantime, Resume Build CI: https://ci.nodejs.org/job/node-test-pull-request/19573/

addaleax commented:
@trott I’m not sure whether we should wait for the outcome of the resume build (run stress tests if it comes back green?), but before/after stress tests with a hundred runs or so might be helpful?

gabrielschulhof commented:
I'll move the test from `atexit(3)` to the spawn-child-synchronously-then-check-stderr pattern. Re: https://github.com/nodejs/node/issues/25034#issuecomment-447623645

gabrielschulhof commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/19576/

gabrielschulhof commented:
New CI: https://ci.nodejs.org/job/node-test-pull-request/19577/

Trott commented:
The change seems to have fixed it on AIX but now it fails on Windows.

```console
02:28:58 not ok 554 addons/at-exit/test
02:28:58   ---
02:28:58   duration_ms: 0.292
02:28:58   severity: fail
02:28:58   exitcode: 1
02:28:58   stack: |-
02:28:58     assert.js:86
02:28:58       throw new AssertionError(obj);
02:28:58       ^
02:28:58     
02:28:58     AssertionError [ERR_ASSERTION]: Expected values to be strictly deep-equal:
02:28:58     + actual - expected
02:28:58     
02:28:58       [
02:28:58     +   'at_exit_cb1\r',
02:28:58     +   'at_exit_cb2\r',
02:28:58     +   'at_exit_cb2\r'
02:28:58     -   'at_exit_cb1',
02:28:58     -   'at_exit_cb2',
02:28:58     -   'at_exit_cb2'
02:28:58       ]
02:28:58         at Object.<anonymous> (c:\workspace\node-test-binary-windows\test\addons\at-exit\test.js:20:8)
02:28:58         at Module._compile (internal/modules/cjs/loader.js:718:30)
02:28:58         at Object.Module._extensions..js (internal/modules/cjs/loader.js:729:10)
02:28:58         at Module.load (internal/modules/cjs/loader.js:617:32)
02:28:58         at tryModuleLoad (internal/modules/cjs/loader.js:560:12)
02:28:58         at Function.Module._load (internal/modules/cjs/loader.js:552:3)
02:28:58         at Function.Module.runMain (internal/modules/cjs/loader.js:771:12)
02:28:58         at executeUserCode (internal/bootstrap/node.js:318:15)
02:28:58         at startExecution (internal/bootstrap/node.js:252:3)
02:28:58         at startup (internal/bootstrap/node.js:203:3)
02:28:58   ...
```

gabrielschulhof commented:
New CI: https://ci.nodejs.org/job/node-test-pull-request/19588/

Trott commented on test/addons/at-exit/test.js:
```diff
@@ -20,7 +20,7 @@ assert.ifError(p.error);
   p.stderr
     .toString()
-    .split('\n')
+    .split(/\n|\r\n?/g)
```
NIt:

```suggestion
    .split(/\r?\n/g)
```

Trott replied:
Or perhaps even better:

```suggestion
    .split(os.EOL)
```

gabrielschulhof commented:
New CI with `EOL` constant: https://ci.nodejs.org/job/node-test-pull-request/19589/

gabrielschulhof commented:
@addaleax @joyeecheung I changed the `at-exit` addon test to spawn a child process and analyse its stderr instead of using a libc `atexit(3)` hook, because the hook would be unloaded before process exit and AIX was (rightly perhaps) segfaulting because of that.

addaleax commented on test/addons/at-exit/binding.cc:
```diff
@@ -10,33 +10,25 @@ using v8::Local;
   AtExit(at_exit_cb1, exports->GetIsolate());
   AtExit(at_exit_cb2, cookie);
   AtExit(at_exit_cb2, cookie);
-  atexit(sanity_check);
```
Would a static destructor also solve the problem? That would run at unload time

gabrielschulhof replied:
I think this is the best alternative.

Trott commented:
CI failure seems build/infra related. Resume Build CI: https://ci.nodejs.org/job/node-test-pull-request/19591/

gabrielschulhof commented:
@addaleax I switched to a library destructor. Hope I got the Windows magic right ... PTAL!

CI: https://ci.nodejs.org/job/node-test-pull-request/19631/ – still pending so the link might be broken at first.

Trott commented:
@addaleax @joyeecheung Changed test looks good to you? Is this author-ready and landable?

gabrielschulhof commented:
Landed in 622e348d8f70a4ec006ee1ce9207a6a5bc3fc325 followed by 13abc6adfb9f6a53618ca7d533d31b6a5d26dcec.

MylesBorins commented:
This change broke the `node-report` module

It seems that is was somehow related to nan? 

https://github.com/nodejs/node-report/issues/116

Fix landed in nan and awaiting a release https://github.com/nodejs/nan/pull/831

Would someone be willing to open a backport so we can track the progress?

addaleax commented:
> This change broke the `node-report` module

@MylesBorins This seems very unlikely … can you double-check and maybe share how you tested this?

> Fix landed in nan and awaiting a release nodejs/nan#831

That PR was opened before this one landed and the contents seem very much unrelated…

BridgeAR commented:
This has been backported and released in v11.7.0 but it turned out that this actually breaks node-report. Therefore I am going to remove the backport again and release v11.7.1 without this.
Please have another look at this so it does not cause any issues with node-report.

addaleax commented:
I’m sorry that I did not add a test case for the issue we’re seeing with node-report when I wrote the original PR at https://github.com/nodejs/node/pull/23319 – I’ve opened https://github.com/nodejs/node/pull/25577 to address it.

(This does not mean that this is not a bug in `node-report`, btw – it is. But there’s a lot of addons that run into similar issues with this PR, I assume.)

richardlau commented:
> (This does not mean that this is not a bug in `node-report`, btw – it is. But there’s a lot of addons that run into similar issues with this PR, I assume.)

If it's something we can improve, please report/submit a PR over in https://github.com/nodejs/node-report. Note that one of the issues with addons and `master` is that `nan` requires an update to work with the version of V8 in `master` (AFAIK https://github.com/nodejs/nan/pull/831 has yet to be released) which means addons currently fail in citgm, which is obviously masking stuff like this.

addaleax commented:
@richardlau I can try to do that, but given the work on integrating it into core in #22712, I don’t think I’d personally want to spend much time on the npm module… the first problem I encountered here is that `nodereport_trigger_async` is never removed from the event loop (e.g. in an Environment cleanup hook). Generally, in #22712 we had (and I still partly have) a hard time figuring out what lifetime management objects from node-report should have and how much global state is actually desirable and managable.

(Also, it’s absolutely understandable if you’re not interested in going through the nearly one thousand comments in https://github.com/nodejs/node/pull/22712, but there’s a lot in there that applies to the npm module as well, including this issue iirc.)
